Lib/subprocess.py

```diff
@@ -968,12 +972,40 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
             # Process startup details
             if startupinfo is None:
                 startupinfo = STARTUPINFO()
-            if -1 not in (p2cread, c2pwrite, errwrite):
+
+            use_std_handles = -1 not in (p2cread, c2pwrite, errwrite)
+            if use_std_handles:
                 startupinfo.dwFlags |= _winapi.STARTF_USESTDHANDLES
                 startupinfo.hStdInput = p2cread
                 startupinfo.hStdOutput = c2pwrite
                 startupinfo.hStdError = errwrite
 
+            attribute_list = startupinfo.lpAttributeList
+            have_handle_list = (attribute_list and
+                                "handle_list" in attribute_list and
+                                attribute_list["handle_list"])
+
+            # If we were given an handle_list or need to create one
+            if have_handle_list or (use_std_handles and close_fds):
+                if attribute_list is None:
+                    attribute_list = startupinfo.lpAttributeList = {}
+                handle_list = attribute_list.setdefault("handle_list", [])
+
+                if use_std_handles:
+                    handle_list += [int(p2cread), int(c2pwrite), int(errwrite)]
+
+                handle_list[:] = self._filter_handle_list(handle_list)
+
+                if handle_list:
+                    if have_handle_list and not close_fds:
```

Comment:
You can get here with `have_handle_list = False` by not passing an handle_list explicitly and requesting stdio redirection.

```python
import subprocess
subprocess.Popen(['notepad.exe'], stdout=subprocess.PIPE)
```

The condition on [line 989](https://github.com/python/cpython/pull/1218/files/af4e1d59213b729e9a5d1c3be59c7b948888a203#diff-cc136486b4a8e112e64b57436a0619ebR989) is for either `have_handle_list` or having requested stdio redirection with `close_fds`.

`have_handle_list` really means that the user has explicitly passed in a non empty handle_list. I placed the check for the warning there instead of someplace else, that might have been clearer, so that we will only give the warning if we actually override `close_fds`. Maybe a different name for `have_handle_list` will make this clearer? If you can come up with one...